sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -0,0 +1,594 @@
+      s"if (shouldStop()) { $idx = $rowidx + 1; return; }"
+    } else {
+      "// shouldStop check is eliminated"
+    }
+    s"""
+       |if ($batch == null) {
+       |  $nextBatchFuncName();
+       |}
+       |while ($batch != null) {
+       |  int $numRows = $batch.numRows();
+       |  int $localEnd = $numRows - $idx;
+       |  for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
+       |    int $rowidx = $idx + $localIdx;
+       |    ${consume(ctx, columnsBatchInput).trim}
+       |    $shouldStop
+       |  }
+       |  $idx = $numRows;
+       |  $batch = null;
+       |  $nextBatchFuncName();
+       |}
+       |$scanTimeMetric.add($scanTimeTotalNs / (1000 * 1000));
+       |$scanTimeTotalNs = 0;
+     """.stripMargin
+  }
+
+  override def inputRDDs(): Seq[RDD[InternalRow]] = {
+    child.asInstanceOf[CodegenSupport].inputRDDs()
+  }
+}
+
+/**
+ * Provides an optimized set of APIs to append row based data to an array of
+ * [[WritableColumnVector]].
+ */
+private[execution] class RowToColumnConverter(schema: StructType) extends Serializable {
+  private val converters = schema.fields.map {
+    f => RowToColumnConverter.getConverterForType(f.dataType)
+  }
+
+  final def convert(row: InternalRow, vectors: Array[WritableColumnVector]): Unit = {
+    var idx = 0
+    while (idx < row.numFields) {
+      converters(idx).append(row, idx, vectors(idx))
+      idx += 1
+    }
+  }
+}
+
+/**
+ * Provides an optimized set of APIs to extract a column from a row and append it to a
+ * [[WritableColumnVector]].
+ */
+private object RowToColumnConverter {
+  private abstract class TypeConverter extends Serializable {
+    def append(row: SpecializedGetters, column: Int, cv: WritableColumnVector): Unit
+  }
+
+  private def getConverterForType(dataType: DataType): TypeConverter = {
+    dataType match {
+      case BooleanType => BooleanConverter
+      case ByteType => ByteConverter
+      case ShortType => ShortConverter
+      case IntegerType => IntConverter
+      case FloatType => FloatConverter
+      case LongType => LongConverter
+      case DoubleType => DoubleConverter
+      case DateType => IntConverter
+      case TimestampType => LongConverter
+      case StringType => StringConverter
+      case CalendarIntervalType => CalendarConverter
+      case at: ArrayType => new ArrayConverter(getConverterForType(at.elementType))
+      case st: StructType => new StructConverter(st.fields.map(
+        (f) => getConverterForType(f.dataType)))
+      case dt: DecimalType => new DecimalConverter(dt)
+      case mt: MapType => new MapConverter(getConverterForType(mt.keyType),
+        getConverterForType(mt.valueType))
+      case unknown => throw new UnsupportedOperationException(
+        s"Type $unknown not supported")
+    }
+  }
+
+  private object BooleanConverter extends TypeConverter {
+    override def append(row: SpecializedGetters, column: Int, cv: WritableColumnVector): Unit =
+      if (row.isNullAt(column)) {
```

Comment:
Could `TypeConverter` take care of writing the nulls so it's not duplicated on each converter?

Reply:
Possibly.  Structs need a different API to append a null, so I would have to special case those within the parent code.  I also was thinking of taking nullability into account to avoid some of the conditionals all together, but yes, I think there is the possibility of reducing the amount of code further.  I'll see what I can come up with.